compiler/shared_ast/interpreter.ml

```diff
@@ -0,0 +1,652 @@
+    match List.length excepts - empty_count with
+    | 0 -> (
+      let just = evaluate_expr ctx just in
+      match Marked.unmark just with
+      | EEmptyError -> Marked.mark m EEmptyError
+      | ELit (LBool true) -> evaluate_expr ctx cons
+      | ELit (LBool false) -> Marked.same_mark_as EEmptyError e
+      | _ ->
+        Errors.raise_spanned_error (Expr.pos e)
+          "Default justification has not been reduced to a boolean at \
+           evaluation (should not happen if the term was well-typed")
+    | 1 -> List.find (fun sub -> not (is_empty_error sub)) excepts
+    | _ ->
+      Errors.raise_multispanned_error
+        (List.map
+           (fun except ->
+             Some "This consequence has a valid justification:", Expr.pos except)
+           (List.filter (fun sub -> not (is_empty_error sub)) excepts))
+        "There is a conflict between multiple valid consequences for assigning \
+         the same variable.")
+  | EIfThenElse { cond; etrue; efalse } -> (
+    propagate_empty_error (evaluate_expr ctx cond)
+    @@ fun cond ->
+    match Marked.unmark cond with
+    | ELit (LBool true) -> evaluate_expr ctx etrue
+    | ELit (LBool false) -> evaluate_expr ctx efalse
+    | _ ->
+      Errors.raise_spanned_error (Expr.pos cond)
+        "Expected a boolean literal for the result of this condition (should \
+         not happen if the term was well-typed)")
+  | EArray es ->
+    propagate_empty_error_list (List.map (evaluate_expr ctx) es)
+    @@ fun es -> Marked.mark m (EArray es)
+  | EEmptyError -> Marked.same_mark_as EEmptyError e
+  | EErrorOnEmpty e' -> (
+    match evaluate_expr ctx e' with
+    | EEmptyError, _ ->
+      Errors.raise_spanned_error (Expr.pos e')
+        "This variable evaluated to an empty term (no rule that defined it \
+         applied in this situation)"
+    | e -> e)
+  | EAssert e' ->
+    propagate_empty_error (evaluate_expr ctx e') (fun e ->
+        match Marked.unmark e with
+        | ELit (LBool true) -> Marked.mark m (ELit LUnit)
+        | ELit (LBool false) -> (
+          match Marked.unmark (Expr.skip_wrappers e') with
+          | EApp
+              {
+                f = EOp { op; _ }, _;
+                args = [((ELit _, _) as e1); ((ELit _, _) as e2)];
+              } ->
+            Errors.raise_spanned_error (Expr.pos e')
+              "Assertion failed: %a %a %a"
+              (Expr.format ctx ~debug:false)
+              e1 Print.operator op
+              (Expr.format ctx ~debug:false)
+              e2
+          | _ ->
+            Cli.debug_format "%a" (Expr.format ctx) e';
+            Errors.raise_spanned_error (Expr.mark_pos m) "Assertion failed")
+        | _ ->
+          Errors.raise_spanned_error (Expr.pos e')
+            "Expected a boolean literal for the result of this assertion \
+             (should not happen if the term was well-typed)")
+
+let interpret_program_lcalc p s : (Uid.MarkedString.info * ('a, 'm) gexpr) list
```

Comment:
Do you really have to duplicate the code for `interpret_program_lcalc` and `interpret_program_dcalc` ?

Reply:
yes there is typing issues. From the discussion with @AltGr, we concluded we need two differents functions.